### ISSUE TITLE
staruml: unable to load old files, broken devtools, locale warnings

### ISSUE THREAD

**Issue by steveeJ** [NixOS/nixpkgs#40585]
## Issue description
I'm encountering more issues with staruml, possibly introduced by
https://github.com/NixOS/nixpkgs/pull/40551.

### Locale warning

```
(process:18045): Gtk-WARNING **: 11:26:46.762: Locale not supported by C library.
	Using the fallback 'C' locale.
```

My setup:

```
$ locale                 
LANG=en_US.UTF-8
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

### devtools feature is missing glibc
When trying to activate the developer tools from *Debug -> Show Devools*, the following error appears in the terminal:

```
/nix/store/h8y6llgmi366vi6z6zx1yh84mb0ghd80-chromium-66.0.3359.139/libexec/chromium/chromium: /nix/store/2kcrj1ksd2a14bm5sky182fv2xwfhfap-glibc-2.26-131/lib/libc.so.6: version `GLIBC_2.27' not found (required by /nix/store/izpra5d6slpacbq8mgx5vanz5axv9fny-systemd-238-lib/lib/libsystemd.so.0)
```

### Opening previous diagrams fails
Unfortunately I haven't yet found a real indicator for why this is failing. 

### Steps to reproduce
Install staruml from https://github.com/NixOS/nixpkgs/commit/4c2913215c9b1139a2d4a1afe1f831c8bcc06b4f.

**Comment by johnae** [comment]
I have the same or similar error when trying to load libsystemd.so dynamically. This problem wasn't present when using libsystemd 237. This is what I see:

```
ib/systemd-journal.moon:3: /nix/store/2kcrj1ksd2a14bm5sky182fv2xwfhfap-glibc-2.26-131/lib/libc.so.6: version `GLIBC_2.27' not found (required by /nix/store/izpra5d6slpacbq8mgx5vanz5axv9fny-systemd-238-lib/lib/libsystemd.so)
```

**Comment by johnae** [comment]
Actually my issue may have been resolved by reinstalling the custom package I was using - I believe it depended on an older version of glibc and so it was already in memory when libsystemd was loaded but ofc depending on a later glibc.

**Comment by steveeJ** [comment]
> I believe it depended on an older version of glibc and so it was already in memory when libsystemd was loaded but ofc depending on a later glibc.

I don't understand your thought process. What do you mean by it already being in memory? Any new executable you create from updating the dependencies of your derivation will have a unique store path, and shouldn't try to load any unreferenced libraries. The RPATH and LD_LIBRARY_PATH should be unrelated to other versions of the same application.

**Comment by johnae** [comment]
Ok well, since luajit (which my software compiles into the executable) depends on glibc and was compiled against a certain path and version, that libc would be in memory when I run my executable. 

Since there is an ffi mechanism in luajit and in this case I dynamically load libsystemd into my program I believe the issue for me was that an older libc was already loaded before ffi came along and loaded libsystemd which then looked at some version constant of libc (I presume from the error) and determined that it didn't like the libc version already used by my executable, so it refused to continue.

I may be wrong here ofc it's just that when I rebuilt my custom package it solved the problem.

The spec does sound like this may be the issue: http://pubs.opengroup.org/onlinepubs/9699919799/functions/dlopen.html

**Comment by steveeJ** [comment]
@johnae I'm still skeptical if what you're thinking about is related to the problem we're facing. Every process has its own virtual address space so calls to `dlopen()` shouldn't interfere with each other if they're from two different processes.
It is possible that long-running processes that use dlopen more than once, and the library changes within the current LIB_LIBRARY_PATH, there will be issues similar to what you describe. But this case is also not possible on NixOS since every executable is wrapped to use a read-only LD_LIBRARY_PATH.

**Comment by steveeJ** [comment]
@johnae 
> In my case, the only address space is the one of the process that dynamically loads libsystemd.so

Do you know which application does that in your case? Just to ensure we're on the same page: you can spawn multiple processes from the same application; each having their own address space

**Comment by johnae** [comment]
@steveeJ Yeah as I mentioned earlier it's my own application, I guess you could see it as a pretty straightforward C application that embeds Luajit which includes an ffi implementation through which I dynamically load libsystemd. It does not fork multiple processes (unless actually using the fork syscall ofc).

**Comment by steveeJ** [comment]
@johnae with multiple processes I mean that if you run the same application you usually get two processes.

**Comment by johnae** [comment]
@steveeJ I'm not following. Running the program wouldn't run anything other than that program unless the program does something (like call fork). If you run it again (like in another terminal for example) you'd get another process ofc...
